sycl-jit/jit-compiler/lib/rtc/PostLinkActions.cpp

```diff
@@ -0,0 +1,116 @@
+//==------------------------ PostLinkActions.cpp ---------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PostLinkActions.h"
+
+#include <llvm/IR/Constants.h>
+#include <llvm/SYCLLowerIR/DeviceGlobals.h>
+#include <llvm/Transforms/Utils/GlobalStatus.h>
+
+using namespace llvm;
+
+bool jit_compiler::post_link::removeSYCLKernelsConstRefArray(Module &M) {
+  GlobalVariable *GV = M.getGlobalVariable("llvm.used");
+
+  if (!GV) {
+    return false;
+  }
+  assert(GV->user_empty() && "Unexpected llvm.used users");
+  Constant *Initializer = GV->getInitializer();
+  GV->setInitializer(nullptr);
+  GV->eraseFromParent();
+
+  // Destroy the initializer and all operands of it.
+  SmallVector<Constant *, 8> IOperands;
+  for (auto It = Initializer->op_begin(); It != Initializer->op_end(); It++)
+    IOperands.push_back(cast<Constant>(*It));
+  assert(llvm::isSafeToDestroyConstant(Initializer) &&
+         "Cannot remove initializer of llvm.used global");
+  Initializer->destroyConstant();
+  for (auto It = IOperands.begin(); It != IOperands.end(); It++) {
+    auto Op = (*It)->stripPointerCasts();
+    auto *F = dyn_cast<Function>(Op);
+    if (llvm::isSafeToDestroyConstant(*It)) {
+      (*It)->destroyConstant();
+    } else if (F && F->getCallingConv() == CallingConv::SPIR_KERNEL &&
+               !F->use_empty()) {
+      // The element in "llvm.used" array has other users. That is Ok for
+      // specialization constants, but is wrong for kernels.
+      llvm::report_fatal_error("Unexpected usage of SYCL kernel");
```

Comment:
Do we ever expect to hit this during user code compilation? If yes, we should probably fail more gracefully and report back through a failed `RTCResult`.

Reply:
Good point, however, I wasn't able to trigger the generation of `llvm.used` and `llvm.compiler.used` with the feature set we currently support. Hence I'm proposing to just assert that they're not present, and to drop the copied code from sycl-post-link for now.